# script to download data from TCGA using TCGAbiolinks

#TCGAbiolinks: Searching GDC database
#https://www.youtube.com/redirect?event=video_description&redir_token=QUFFLUhqbFo1T0ZMdVlKeDBYTTVUOUJQNDBkdFpiVkZCZ3xBQ3Jtc0ttRnBaQzdQeTFEVXZaWGZiRnRmVFRzWHBUZk1aV0hrYll0bnUzT2t3WHhFcE1saW1fNy1PTDc3NTByWVZvb09LT1dCVElfMzYxN2VpeTk0MEZLQ1hGMlRpMDhrb3F6RDBCTUNpWHVFMzJ1cTdKYTBLOA&q=https%3A%2F%2Fbioconductor.org%2Fpackages%2Frelease%2Fbioc%2Fvignettes%2FTCGAbiolinks%2Finst%2Fdoc%2Fquery.html&v=UWXv9dUpxNE

BiocManager::install(c("sesame","SummarizedExperiment","dplyr"))

setwd("/shared/ifbstor1/projects/cible_cd51/R_projects")


library(sesame)
library(TCGAbiolinks)
library(tidyverse)
library(maftools)
library(pheatmap)
library(SummarizedExperiment)


# get a list of projects
gdcprojects <- getGDCprojects()

# another function to filter projects : kind of data is available and how many cases
getProjectSummary('TCGA-BRCA')



# building a query
query_TCGA <- GDCquery(project = 'TCGA-BRCA',
                       data.category = 'Transcriptome Profiling')
output_query_TCGA <- getResults(query_TCGA) # so many files and several cases with acceses and generated by different techniques (mirna, ...) 


# build a query to retrieve gene EXPRESSION data ------------
query_TCGA <- GDCquery(project = 'TCGA-BRCA',
                       data.category = 'Transcriptome Profiling',
                       experimental.strategy = 'RNA-Seq',
                       workflow.type = 'STAR - Counts',
                       access = 'open', # because i do not have autorization to get files
                       barcode = c('TCGA-LL-A73Y-01A-11R-A33J-07',
                                   'TCGA-E2-A1IU-01A-11R-A14D-07',
                                   'TCGA-AO-A03U-01B-21R-A10J-07')) # download like 3 files

getResults(query_TCGA)

# download data - GDCdownload (this will create a folder in working dir)
GDCdownload(query_TCGA)


# prepare data : extract data from this files
  <- GDCprepare(query_TCGA, summarizedExperiment = TRUE)

# create a matrix with patient in columns and genes in rows
brca_matrix <- assay(tcga_brca_data, 'fpkm_unstrand')


# build a query to retrieve DNA METHYLATION data --------------
query_methly <- GDCquery(project = 'TCGA-GBM', # help to create folder with the same name as the project
                         data.category = 'DNA Methylation',
                         platform = 'Illumina Human Methylation 27', # only with DNA meth data
                         access = 'open',
                         data.type = 'Methylation Beta Value',
                         barcode = c('TCGA-19-0962-01B-01D-0521-05', 'TCGA-06-0137-01A-01D-0218-05'))

output_query_methyl <- getResults(query_methly)

GDCdownload(query_methly)


# plot probes showing differences in beta values between samples                                               
dna.meth <- GDCprepare(query_methly, summarizedExperiment = TRUE)
assay(dna.meth) # ACCESS TO methylation values

idx <- dna.meth %>% 
  assay %>% # access to meth value
  rowVars() %>%  # calculate variance across samples
  order(decreasing = TRUE) %>%  # order sample according to variance
  head(10) # we need the top 10 samples

idx

# plot
pheatmap(assay(dna.meth)[idx,], main = 'Top 10 samples with highest methylation variance')


# download and visualize MUTATION data from TCGA ----------------------
query_mutation <- GDCquery(project = 'TCGA-BRCA',
                           data.category = 'Simple Nucleotide Variation', 
                           access = 'open',
                           barcode = c('TCGA-LL-A73Y-01A-11D-A33E-09,TCGA-LL-A73Y-10B-01D-A33H-09',
                                       'TCGA-E9-A1NH-01A-11D-A14G-09,TCGA-E9-A1NH-11A-33D-A14G-09'))

output_query_mutation <- getResults(query_mutation)

GDCdownload(query_mutation)

maf <- GDCprepare(query_mutation, summarizedExperiment = TRUE)

# maftools utils to read and create dashboard
maftools.input <- read.maf(maf)

plotmafSummary(maf = maftools.input,
               addStat = 'median',
               rmOutlier = TRUE,
               dashboard = TRUE)

# oncoprint
oncoplot(maf = maftools.input,
         top = 10,
         removeNonMutated = TRUE)





